### ISSUE TITLE
KillSwitch: user has no permission to perform this operation

### ISSUE THREAD

**Issue by HacKanCuBa** [ProtonVPN/protonvpn-nm-lib#7]
This issue is caused by enabling the KillSwitch feature, and is related to NetworkManager. Originally reported in the `linux-cli` repo, but it belongs here since it's caused by this lib: https://github.com/ProtonVPN/linux-cli/issues/4

```
Setting up ProtonVPN.
Connecting to ProtonVPN on [REDACTED] with UDP.
ERROR:dbus.connection:Exception in handler for D-Bus signal:
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/dbus/connection.py", line 232, in maybe_handle_message
    self._handler(*args, **kwargs)
  File "/usr/lib/python3/dist-packages/protonvpn_nm_lib/core/connection_backend/nm_client/monitor_vpn_connection_start.py", line 58, in on_vpn_state_changed
    env.killswitch.manage(KillSwitchActionEnum.SOFT)
  File "/usr/lib/python3/dist-packages/protonvpn_nm_lib/core/killswitch/killswitch.py", line has no permission to perform this operation82, in manage
    actions_dict = {
  File "/usr/lib/python3/dist-packages/protonvpn_nm_lib/core/killswitch/killswitch.py", line 280, in setup_soft_connection
    self.setup_post_connection_ks(None, activating_soft_connection=True)
  File "/usr/lib/python3/dist-packages/protonvpn_nm_lib/core/killswitch/killswitch.py", line 221, in setup_post_connection_ks
    self.activate_connection(self.ks_conn_name)
  File "/usr/lib/python3/dist-packages/protonvpn_nm_lib/core/killswitch/killswitch.py", line 404, in activate_connection
    conn_dict = self.nm_wrapper.search_for_connection( # noqa
  File "/usr/lib/python3/dist-packages/protonvpn_nm_lib/core/dbus/dbus_network_manager_wrapper.py", line 56, in search_for_connection
    all_connection_properties = self.get_settings_from_connection(
  File "/usr/lib/python3/dist-packages/protonvpn_nm_lib/core/dbus/dbus_network_manager_wrapper.py", line 418, in get_settings_from_connection
    return iface.GetSettings()
  File "/usr/lib/python3/dist-packages/dbus/proxies.py", line 72, in __call__
    return self._proxy_method(*args, **keywords)
  File "/usr/lib/python3/dist-packages/dbus/proxies.py", line 141, in __call__
    return self._connection.call_blocking(self._named_service,
  File "/usr/lib/python3/dist-packages/dbus/connection.py", line 652, in call_blocking
    reply_message = self.send_message_with_reply_and_block(
dbus.exceptions.DBusException: org.freedesktop.NetworkManager.Settings.PermissionDenied: uid 1000 has no permission to perform this operation
```

This happens in Debian 11 for me, and according to the mentioned issue, it also occurs in Manjaron, Ubuntu 21.04 and OpenSuse.

It seems relates to Polkit rules, the lack of one specifically. The lib should check for permission, or try and catch the exception to exit and fail properly. It should ideally also provide some instructions on how to solve it: can we create a polkit rule? can we execute that action in a different way?

**Comment by calexandru2018** [comment]
Hey @HacKanCuBa 

Appreciate the report. Ideally polkit rules are correctly configured and shouldn't me messed with. Can you check if the dummy kernel module is installed on your system ?

**Comment by HacKanCuBa** [comment]
Sure thing! Which kernel module? The one named `dummy`? If so, it is loaded:

```
:~$ sudo modinfo dummy
filename:       /lib/modules/5.10.0-8-amd64/kernel/drivers/net/dummy.ko
alias:          rtnl-link-dummy
license:        GPL
depends:        
retpoline:      Y
intree:         Y
name:           dummy
vermagic:       5.10.0-8-amd64 SMP mod_unload modversions 
sig_id:         PKCS#7
signer:         Debian Secure Boot CA
sig_key:        4B:6E:F5:AB:CA:66:98:25:17:8E:05:2C:84:66:7C:CB:C0:53:1F:8C
sig_hashalgo:   sha256
signature:      87:46:8E:C6:84:99:5C:31:6F:A5:96:29:3D:4D:17:9D:37:D6:23:1B:
		6F:D5:B2:1F:88:E4:00:49:7E:A8:B0:86:4B:13:F2:7A:FD:D6:60:31:
		05:EB:3A:6E:5D:A3:77:11:B8:95:73:FD:36:F3:06:3B:37:BC:18:FD:
		C9:0C:D7:61:66:C0:23:CE:71:A0:A4:BA:B4:DB:9E:49:5A:23:60:8F:
		80:8C:67:E6:93:15:15:1D:9A:16:C1:65:0D:1F:1C:0C:2F:A6:6B:03:
		F2:CD:85:05:39:87:84:B9:30:8F:C3:9E:22:11:E2:CD:6E:EA:14:3A:
		60:7F:59:F9:FB:39:ED:F4:68:6A:4E:B9:AC:A7:3E:E0:1A:2C:EC:D2:
		08:A9:EB:20:D4:02:15:55:87:A0:0E:83:80:C2:DB:79:49:92:B6:59:
		BC:EE:B7:90:B5:91:C8:20:0D:C9:02:9C:94:5B:52:10:FC:F2:ED:24:
		E8:E4:69:02:14:29:6F:B5:30:BD:FC:23:3A:8F:5E:8A:AD:28:E7:B1:
		4C:E2:60:F5:8B:72:8B:06:77:50:70:D6:3F:C7:86:70:88:A0:A8:3A:
		BA:C0:0A:10:44:FD:06:6F:34:E5:80:D4:35:E7:03:91:68:96:E0:23:
		83:E3:54:3D:2A:16:47:81:FF:59:35:4C:BA:1F:D4:3F
parm:           numdummies:Number of dummy pseudo devices (int)
```

**Comment by calexandru2018** [comment]
It seems indeed. Could you check your polkit rules ? Should be similar to [these](https://github.com/ProtonVPN/linux-cli/issues/42#issuecomment-889178158)

**Comment by HacKanCuBa** [comment]
Here they are: [org.freedesktop.NetworkManager.policy.txt](https://github.com/ProtonVPN/protonvpn-nm-lib/files/6919868/org.freedesktop.NetworkManager.policy.txt)

**Comment by HacKanCuBa** [comment]
`diff yours mine`:
```
168c168
<     <description xml:lang="pl">Umieszczenie usługi NetworkManager w stanie uśpienia lub przebudzenie jej (powinno być używane tylko przez systemowe zarządzanie zasilaniem)</description>
---
>     <description xml:lang="pl">Umieszczenie usługi NetworkManager w stanie uśpienia lub przebudzenie jej (powinno być używane tylko przez systemowe zarządzanie zasilaniem)</description>
208c208
<     <message xml:lang="pl">Ustawienia systemu uniemożliwiają umieszczenie usługi NetworkManager w stanie uśpienia lub przebudzenie jej</message>
---
>     <message xml:lang="pl">Ustawienia systemu uniemożliwiają umieszczenie usługi NetworkManager w stanie uśpienia lub przebudzenie jej</message>
947,948c947,948
<       <allow_inactive>yes</allow_inactive>
<       <allow_active>yes</allow_active>
---
>       <allow_inactive>auth_admin_keep</allow_inactive>
>       <allow_active>auth_admin_keep</allow_active>
1206c1206
< </policyconfig>
---
> </policyconfig>
\ No newline at end of file
```

Nothing significant...

**Comment by oscarvarto** [comment]
@calexandru2018 I am also hitting this problem. I am on Manjaro. By any chance, do you know how could I install/load the dummy kernel module mentioned above?

**Comment by tschaffter** [comment]
I have the same error as the one reported in the initial comment in this issue. I'm running a fresh install of Ubuntu 20.04.

**Comment by Syntrice** [comment]
Had this same problem for months, really need this to be fixed and I have no idea how to get it to work myself.

**Comment by df71** [comment]
> `diff yours mine`:
> 
> ```
> 168c168
> <     <description xml:lang="pl">Umieszczenie usługi NetworkManager w stanie uśpienia lub przebudzenie jej (powinno być używane tylko przez systemowe zarządzanie zasilaniem)</description>
> ---
> >     <description xml:lang="pl">Umieszczenie usługi NetworkManager w stanie uśpienia lub przebudzenie jej (powinno być używane tylko przez systemowe zarządzanie zasilaniem)</description>
> 208c208
> <     <message xml:lang="pl">Ustawienia systemu uniemożliwiają umieszczenie usługi NetworkManager w stanie uśpienia lub przebudzenie jej</message>
> ---
> >     <message xml:lang="pl">Ustawienia systemu uniemożliwiają umieszczenie usługi NetworkManager w stanie uśpienia lub przebudzenie jej</message>
> 947,948c947,948
> <       <allow_inactive>yes</allow_inactive>
> <       <allow_active>yes</allow_active>
> ---
> >       <allow_inactive>auth_admin_keep</allow_inactive>
> >       <allow_active>auth_admin_keep</allow_active>
> 1206c1206
> < </policyconfig>
> ---
> > </policyconfig>
> \ No newline at end of file
> ```
> 
> Nothing significant...

Same issue for me on Ubuntu 20.04 LTS. Also tried changing the <allow_inactive> and <allow_active> values to the ones of the sample polkit rules of @calexandru2018 but problem remains.

**Comment by jr1221** [comment]
Same issue here... A CLI app should be able to run without NetworkManager.  There seems to be little support for linux out of ProtonVPN, its too bad.

**Comment by mike-lloyd03** [comment]
Been trying for two days to make this work. It seems there is no way to connect to the VPN on Arch because of this NetworkManager permissions issue. Has anyone been successful with this?

```
2022-04-01 00:35:41,594 — cli.py — INFO — __init__:29 — ProtonVPN CLI v3.11.1 (protonvpn-nm-lib v3.8.0; proton-client v0.7.1)
2022-04-01 00:35:41,601 — cli.py — INFO — __init__:64 — CLI command: Namespace(command='c', version=False, help=False, get_logs=False)
2022-04-01 00:35:41,604 — cli.py — INFO — connect:128 — Options: Namespace(servername=None, fastest=True, random=False, cc=None, sc=False, p2p=False, tor=False, protocol=None, help=False)
2022-04-01 00:35:43,954 — cli_wrapper.py — ERROR — connect:379 — Unable to add IPv6 leak protection connection/interface
Traceback (most recent call last):                                                                        File "/usr/lib/python3.10/site-packages/protonvpn_cli/cli_wrapper.py", line 277, in connect
    self.protonvpn.setup_connection(                                                                      File "/usr/lib/python3.10/site-packages/protonvpn_nm_lib/api.py", line 153, in setup_connection
    self._env.connection_backend.setup(**data)
  File "/usr/lib/python3.10/site-packages/protonvpn_nm_lib/core/connection_backend/nm_client/nm_client.py", line 81, in setup
    self._pre_setup_connection(kwargs.get("entry_ip"))
  File "/usr/lib/python3.10/site-packages/protonvpn_nm_lib/core/connection_backend/nm_client/nm_client.py", line 233, in _pre_setup_connection
    ipv6_lp.manage(KillSwitchActionEnum.ENABLE)                                                           File "/usr/lib/python3.10/site-packages/protonvpn_nm_lib/core/killswitch/ipv6_leak_protection.py", line 62, in manage
    self.add_leak_protection()
  File "/usr/lib/python3.10/site-packages/protonvpn_nm_lib/core/killswitch/ipv6_leak_protection.py", line 100, in add_leak_protection
    self.run_subprocess(
  File "/usr/lib/python3.10/site-packages/protonvpn_nm_lib/core/killswitch/ipv6_leak_protection.py", line 178, in run_subprocess                                                                                    raise exception(exception_msg)
protonvpn_nm_lib.exceptions.EnableIPv6LeakProtectionError: Unable to add IPv6 leak protection connection/interface
```

**Comment by jr1221** [comment]
I gave up.  Maybe I'm not educated in the intricacies of VPNs, but this app seems to be poorly designed, requiring GUI libraries for their cli app is fundamentally flawed.

**Comment by mike-lloyd03** [comment]
Yes. X11 is a silly dependency for for a CLI tool. But as with most things Linux, we can expect better support from the community. 

I installed this tool and it worked on the first try. I wish I saw this last night:

https://github.com/Rafficer/linux-cli-community

**Comment by calexandru2018** [comment]
Hey @Syntrice @df71 @tschaffter @HacKanCuBa @mike-lloyd03 @jr1221 

A [new release of the library](https://github.com/ProtonVPN/protonvpn-nm-lib/releases/tag/3.9.0) has been made which fixes this issue. Could you please update to the latest version ?

**Comment by HacKanCuBa** [comment]
> Hey @Syntrice @df71 @tschaffter @HacKanCuBa @mike-lloyd03 @jr1221
> 
> A [new release of the library](https://github.com/ProtonVPN/protonvpn-nm-lib/releases/tag/3.9.0) has been made which fixes this issue. Could you please update to the latest version ?

I just updated, and it didn't failed. It indicates that it is working, so It seems to be working.
Any tip on how could I put it to test?

**Comment by calexandru2018** [comment]
> > Hey @Syntrice @df71 @tschaffter @HacKanCuBa @mike-lloyd03 @jr1221
> 
> > 
> 
> > A [new release of the library](https://github.com/ProtonVPN/protonvpn-nm-lib/releases/tag/3.9.0) has been made which fixes this issue. Could you please update to the latest version ?
> 
> 
> 
> I just updated, and it didn't failed. It indicates that it is working, so It seems to be working.
> 
> Any tip on how could I put it to test?

That's actually the only way to test it tbh. If the error does not occur then it is solved.

The problem was that, apparently some 3rd party connections were not allowing us to get into their settings (thus the permission denied) when we tried to access them. At the beginning we though it could've been some issue with permissions as it was the one that made most sense, but even after tweaking permissions the issue was there.

Finally (after many months of trying to figure out what could be), based on the logs that our users had provided us, we were able to pinpoint this. Basically this issue only occurred in specific situations where NM had other connections that were created by 3rd parties (ie docker or any other software that creates connections).

Took us some time to figure all this out, as we were not able to recreate this issue in-house and also because the Linux team is quite small (solo dev here).
